operators/src/engine/query.rs

```diff
@@ -43,37 +51,96 @@ impl TestDefault for ChunkByteSize {
 pub trait QueryContext: Send + Sync {
     fn chunk_byte_size(&self) -> ChunkByteSize;
     fn thread_pool(&self) -> &Arc<ThreadPool>;
+
+    fn abort_registration(&self) -> &QueryAbortRegistration;
+    fn abort_trigger(&mut self) -> Result<QueryAbortTrigger>;
+}
+
+pub struct QueryAbortRegistration {
+    valve: Valve,
+}
+
+impl QueryAbortRegistration {
+    pub fn new() -> (Self, QueryAbortTrigger) {
+        let (trigger, valve) = Valve::new();
+
+        (Self { valve }, QueryAbortTrigger { trigger })
+    }
+
+    pub fn wrap<S: Stream>(&self, stream: S) -> QueryAbortWrapper<S> {
+        QueryAbortWrapper {
+            valved: self.valve.wrap(stream),
+        }
+    }
+}
+
+#[pin_project(project = AbortWrapperProjection)]
+pub struct QueryAbortWrapper<S> {
```

Comment:
Could you briefly describe what this is?

Reply:
8e9bbd185e7e2a422a5f9110eed35702276aef52